services/time/src/lib.rs

```diff
@@ -111,62 +113,100 @@ transactions! {
     }
 }
 
-impl Transaction for TxTime {
-    fn verify(&self) -> bool {
-        self.verify_signature(self.pub_key())
-    }
+#[derive(Debug, Fail)]
+#[repr(u8)]
+enum Error {
+    #[fail(display = "Not authored by a validator")]
+    UnknownSender = 0,
 
-    fn execute(&self, view: &mut Fork) -> ExecutionResult {
-        let validator_keys = Schema::new(&view).actual_configuration().validator_keys;
+    #[fail(display = "The validator time is longer than the proposed one")]
+    ValidatorTimeIsLonger = 1,
+}
 
-        // The transaction must be signed by the validator.
-        let signed = validator_keys.iter().any(|&validator| {
-            validator.service_key == *self.pub_key()
-        });
+impl From<Error> for ExecutionError {
+    fn from(value: Error) -> ExecutionError {
+        ExecutionError::new(value as u8)
+    }
+}
+
+impl TxTime {
+    fn signed_by_validator(&self, snapshot: &Snapshot, key: &PublicKey) -> ExecutionResult {
+        let keys = Schema::new(&snapshot).actual_configuration().validator_keys;
+        let signed = keys.iter().any(|k| k.service_key == *key);
         if !signed {
-            return Ok(());
+            Err(Error::UnknownSender)?
+        } else {
+            Ok(())
         }
+    }
 
-        let mut schema = TimeSchema::new(view);
-        match schema.validators_times().get(self.pub_key()) {
+    fn update_validator_time(
+        &self,
+        fork: &mut Fork,
+        key: &PublicKey,
+        proposed_time: SystemTime,
+    ) -> ExecutionResult {
+        let mut schema = TimeSchema::new(fork);
+        match schema.validators_times().get(key) {
             // The validator time in the storage should be less than in the transaction.
-            Some(storage_time) if storage_time >= self.time() => {
-                return Ok(());
-            }
+            Some(time) if time >= proposed_time => Err(Error::ValidatorTimeIsLonger)?,
             // Write the time for the validator.
             _ => {
-                schema.validators_times_mut().put(
-                    self.pub_key(),
-                    self.time(),
-                )
+                schema.validators_times_mut().put(key, proposed_time);
+                Ok(())
             }
         }
+    }
+}
+
+impl Transaction for TxTime {
+    fn verify(&self) -> bool {
+        self.verify_signature(self.pub_key())
+    }
+
+    fn execute(&self, view: &mut Fork) -> ExecutionResult {
+        // The transaction must be signed by the validator.
+        self.signed_by_validator(view.as_ref(), self.pub_key())?;
+
+        self.update_validator_time(
```

Comment:
Again, if `update_validator_time` is a method, then `pub_key` and `time` can be taken from `self`.